crates/chunk/examples/chunk_merkle_proving.rs

```diff
@@ -0,0 +1,429 @@
+    );
+
+    let mut result: Vec<Boolean> = vec![];
+
+    // For each field element, take the first `bit_size` bits
+    for (i, bit_to_take) in scalar_bit_sizes.iter().enumerate() {
+        let element = elements[i]
+            .to_bits_le(&mut cs.namespace(|| "test"))
+            .unwrap();
+
+        result.extend(element.into_iter().take(*bit_to_take));
+    }
+
+    result
+}
+
+/*****************************************
+ * Circuit
+ *****************************************/
+
+struct MerkleChunkCircuit<F: PrimeField + PrimeFieldBits, C: ChunkStepCircuit<F>, const N: usize> {
+    inner: InnerCircuit<F, C, N>,
+}
+
+impl<F: PrimeField + PrimeFieldBits, C: ChunkStepCircuit<F>, const N: usize>
+    MerkleChunkCircuit<F, C, N>
+{
+    fn new(inputs: &[F], post_processing_step: Option<F>) -> Self {
+        Self {
+            inner: InnerCircuit::new(inputs, post_processing_step).unwrap(),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+enum ChunkCircuitSet<F: PrimeField + PrimeFieldBits, C: ChunkStepCircuit<F>, const N: usize> {
+    IterStep(FoldStepWrapper<F, C, N>),
+    CheckEquality(EqualityCircuit<F>),
+}
+
+impl<F: PrimeField + PrimeFieldBits, C: ChunkStepCircuit<F>, const N: usize> StepCircuit<F>
+    for ChunkCircuitSet<F, C, N>
+{
+    fn arity(&self) -> usize {
+        match self {
+            Self::IterStep(fold_step) => fold_step.inner.arity(),
+            Self::CheckEquality(equality_circuit) => equality_circuit.arity(),
+        }
+    }
+
+    fn circuit_index(&self) -> usize {
+        match self {
+            Self::IterStep(fold_step) => *fold_step.inner.step_nbr(),
+            Self::CheckEquality(equality_circuit) => equality_circuit.circuit_index(),
+        }
+    }
+
+    fn synthesize<CS: ConstraintSystem<F>>(
+        &self,
+        cs: &mut CS,
+        pc: Option<&AllocatedNum<F>>,
+        z: &[AllocatedNum<F>],
+    ) -> Result<(Option<AllocatedNum<F>>, Vec<AllocatedNum<F>>), SynthesisError> {
+        match self {
+            Self::IterStep(fold_step) => fold_step.synthesize(cs, pc, z),
+            Self::CheckEquality(equality_circuit) => equality_circuit.synthesize(cs, pc, z),
+        }
+    }
+}
+
+impl<E1: CurveCycleEquipped, C: ChunkStepCircuit<E1::Scalar>, const N: usize> NonUniformCircuit<E1>
+    for MerkleChunkCircuit<E1::Scalar, C, N>
+{
+    type C1 = ChunkCircuitSet<E1::Scalar, C, N>;
+    type C2 = TrivialSecondaryCircuit<<Dual<E1> as Engine>::Scalar>;
+
+    fn num_circuits(&self) -> usize {
+        self.inner.num_fold_steps() + 1
+    }
+
+    fn primary_circuit(&self, circuit_index: usize) -> Self::C1 {
+        if circuit_index == 2 {
+            Self::C1::CheckEquality(EqualityCircuit::new())
+        } else {
+            if let Some(fold_step) = self.inner.circuits().get(circuit_index) {
+                return Self::C1::IterStep(FoldStepWrapper::new(fold_step.clone()));
+            }
+            panic!("No circuit found for index {}", circuit_index)
+        }
+    }
```

Comment:
This implementation reveals that `primary_circuit` is either wrongly implemented locally here or (as I think I observe from looking around) being misused altogether in this `ChunkStepCircuit` abstraction. Since I haven't had time yet to fully probe this, I will hedge and say that it's remotely conceivable that this is correct but just written in an extremely convoluted way. If that turns out to be the case, I will recommend simplifying.

Reply:
Part of the confusion here flows from the wrong interpretation of `num_circuits` above. That method should return the number of alternative circuits SuperNova must manage when folding. It does not indicate the number of folding steps or anything related to the total size of the computation. It relates to the shape of the proof, which is insensitive to details of the actual computation trace that will be proved.

Reply:
You are 100% right, I got confused on the meaning of the API for `Supernova`. I tried to refactor the examples based on what you said.